src/state/zustand/swapperStore/amountSelectors.ts

```diff
@@ -0,0 +1,394 @@
+import { createSelector } from 'reselect'
+import { TradeAmountInputField } from 'components/Trade/types'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { fromBaseUnit, toBaseUnit } from 'lib/math'
+import type { SwapperState } from 'state/zustand/swapperStore/useSwapperStore'
+
+const selectAssetPriceRatio = createSelector(
+  (state: SwapperState) => state.buyAssetFiatRate,
+  (state: SwapperState) => state.sellAssetFiatRate,
+  (buyAssetFiatRate, sellAssetFiatRate) => {
+    return bnOrZero(buyAssetFiatRate).dividedBy(bnOrZero(sellAssetFiatRate)).toString()
+  },
+)
+
+export const selectSellAmountBeforeFeesBaseUnit = createSelector(
+  (state: SwapperState) => state.action,
+  (state: SwapperState) => state.amount,
+  (state: SwapperState) => state.sellAsset?.precision,
+  (state: SwapperState) => state.sellAssetFiatRate,
+  selectAssetPriceRatio,
+  (action, amount, sellAssetPrecision, sellAssetFiatRate, assetPriceRatio): string => {
+    if (!sellAssetPrecision || !sellAssetFiatRate) return '0'
+    switch (action) {
+      case TradeAmountInputField.BUY_CRYPTO:
+        return toBaseUnit(bnOrZero(amount).times(assetPriceRatio), sellAssetPrecision)
+      case TradeAmountInputField.SELL_CRYPTO:
+        return toBaseUnit(amount, sellAssetPrecision)
+      case TradeAmountInputField.BUY_FIAT:
+      case TradeAmountInputField.SELL_FIAT:
+        return toBaseUnit(bnOrZero(amount).dividedBy(sellAssetFiatRate), sellAssetPrecision)
+      default:
+        return '0'
+    }
+  },
+)
+
+export const selectBuyAmountBeforeFeesBaseUnit = createSelector(
+  (state: SwapperState) => state.action,
+  (state: SwapperState) => state.amount,
+  (state: SwapperState) => state.buyAsset?.precision,
+  (state: SwapperState) => state.buyAssetFiatRate,
+  selectAssetPriceRatio,
+  (action, amount, buyAssetPrecision, buyAssetFiatRate, assetPriceRatio): string => {
+    if (!buyAssetPrecision || !buyAssetFiatRate) return '0'
+    switch (action) {
+      case TradeAmountInputField.BUY_CRYPTO:
+        return toBaseUnit(amount, buyAssetPrecision)
+      case TradeAmountInputField.SELL_CRYPTO:
+        return toBaseUnit(bnOrZero(amount).div(assetPriceRatio), buyAssetPrecision)
+      case TradeAmountInputField.BUY_FIAT:
+      case TradeAmountInputField.SELL_FIAT:
+        return toBaseUnit(bnOrZero(amount).dividedBy(buyAssetFiatRate), buyAssetPrecision)
+      default:
+        return '0'
+    }
+  },
+)
+
+export const selectSellAmountBeforeFeesBuyAsset = createSelector(
+  selectAssetPriceRatio,
+  selectSellAmountBeforeFeesBaseUnit,
+  (state: SwapperState) => state.sellAsset?.precision,
+  (assetPriceRatio, sellAmountBeforeFeesBaseUnit, sellAssetPrecision): string | undefined => {
+    if (!sellAssetPrecision) return undefined
+    return bnOrZero(fromBaseUnit(sellAmountBeforeFeesBaseUnit, sellAssetPrecision))
+      .div(assetPriceRatio)
+      .toString()
+  },
+)
+
+export const selectSellAmountBeforeFeesBuyAssetBaseUnit = createSelector(
```

Comment:
Can we invert the selectors deps e.g use the base unit to calculate the precision amount ? Something something loss of precision

Reply:
Absolutely. Inverted in https://github.com/shapeshift/web/pull/4138/commits/dddbde80b84b26edfea966547554a69f2d10f290.